dll/win32/shell32/CDefView.cpp

```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
+    else
+    {
+        if (FAILED(hr = (m_pShellBrowser ? m_pShellBrowser->GetViewStateStream(STGM_READ, &pS) : E_UNEXPECTED)))
+        {
+        loadfallback:
+            hr = GetDefaultViewStream(STGM_READ, &pS);
+            fallback = true;
+        }
+        if (FAILED(hr) || FAILED(hr = Read(pS, &cvs, sizeof(cvs))))
+            return hr;
+        if (cvs.Signature != cvs.SIG)
+            return HResultFromWin32(ERROR_INVALID_DATA);
+        hrColumns = LoadColumnsStream(cols, pS);
+    }
+    m_FolderSettings.ViewMode = cvs.FolderSettings.ViewMode;
+    m_FolderSettings.fFlags &= ~cvs.VALIDFWF;
+    m_FolderSettings.fFlags |= cvs.FolderSettings.fFlags & cvs.VALIDFWF;
+    if (SUCCEEDED(hrColumns))
+    {
+        BOOL failed = FALSE;
+        if ((m_LoadColumnsList = DPA_Create(cols.Count)) != NULL)
+        {
+            for (UINT i = 0; i < cols.Count; ++i)
+            {
+                failed |= !DPA_SetPtr(m_LoadColumnsList, i, (void*)(SIZE_T)cols.Columns[i]);
+            }
+        }
+        if (failed || !cols.Count)
+        {
+            DPA_Destroy(m_LoadColumnsList);
+            m_LoadColumnsList = NULL;
+        }
+    }
+    m_sortInfo.bLoadedFromViewState = !fallback && m_LoadColumnsList && cvs.SortColId != LISTVIEW_SORT_INFO::UNSPECIFIEDCOLUMN;
+    m_sortInfo.bColumnIsFolderColumn = TRUE;
+    m_sortInfo.Direction = cvs.SortDir > 0 ? 1 : -1;
+    m_sortInfo.ListColumn = cvs.SortColId;
+    return hr;
+}
+
+HRESULT CDefView::SaveViewState(IStream *pStream)
+{
+    int sortcol = MapListColumnToFolderColumn(m_sortInfo.ListColumn);
+    PERSISTCLASSICVIEWSTATE cvs;
+    cvs.SortColId = sortcol >= 0 ? (WORD)sortcol : 0;
+    cvs.SortDir = m_sortInfo.Direction;
+    PERSISTCOLUMNS cols;
+    cols.Signature = PERSISTCOLUMNS::SIG;
+    cols.Count = 0;
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_WIDTH | LVCF_SUBITEM;
+    for (UINT i = 0, j = 0; i < PERSISTCOLUMNS::MAXCOUNT && ListView_GetColumn(m_ListView, j, &lvc); ++j)
+    {
+        HRESULT hr = MapListColumnToFolderColumn(lvc.iSubItem);
+        if (SUCCEEDED(hr))
+        {
+            cols.Columns[i] = MAKELONG(hr, lvc.cx);
+            cols.Count = ++i;
+        }
+    }
+    UINT cbColumns = FIELD_OFFSET(PERSISTCOLUMNS, Columns) + (sizeof(*cols.Columns) * cols.Count);
+    UpdateFolderViewFlags();
+
+    IPropertyBag *pPB;
+    HRESULT hr = S_OK;
+    if (pStream)
+    {
+        pStream->AddRef();
+        goto stream;
+    }
+    hr = IUnknown_QueryServicePropertyBag(m_pShellBrowser, SHGVSPB_FOLDER, IID_PPV_ARG(IPropertyBag, &pPB));
+    if (SUCCEEDED(hr))
+    {
+        UINT u32;
+        GetCurrentViewMode(&u32);
+        hr = SHPropertyBag_WriteDWORD(pPB, L"Mode", u32);
+        SHPropertyBag_WriteDWORD(pPB, L"FFlags", m_FolderSettings.fFlags);
+        SHPropertyBag_WriteDWORD(pPB, L"Sort", cvs.SortColId);
+        SHPropertyBag_WriteDWORD(pPB, L"SortDir", cvs.SortDir);
+        pStream = cols.Count ? SHCreateMemStream((LPBYTE)&cols, cbColumns) : NULL;
```

Comment:
Just to confirm: the SHCreateMemStream adds an initial reference on the pStream?

Reply:
Yes